pandas/core/indexes/numeric.py

```diff
@@ -107,20 +107,22 @@ def _can_hold_na(self) -> bool:
         else:
             return False
 
-    @cache_readonly
+    _engine_types: dict[np.dtype : libindex.IndexEngine] = {
+        np.dtype(np.int8): libindex.Int8Engine,
+        np.dtype(np.int16): libindex.Int16Engine,
+        np.dtype(np.int32): libindex.Int32Engine,
+        np.dtype(np.int64): libindex.Int64Engine,
+        np.dtype(np.uint8): libindex.UInt8Engine,
+        np.dtype(np.uint16): libindex.UInt16Engine,
+        np.dtype(np.uint32): libindex.UInt32Engine,
+        np.dtype(np.uint64): libindex.UInt64Engine,
+        np.dtype(np.float32): libindex.Float32Engine,
+        np.dtype(np.float64): libindex.Float64Engine,
+    }
+
```

Comment:
By having  this as a class variable, we avoid rebuilding the dict in every instantiation.

Reply:
does this affect perf?

Reply:
I little bit:

```python
>>> arr = np.arange(1_000_000)
>>> %timeit pd.Int64Index(arr)._engine_type
6.72 ms ± 83.8 µs per loop  # master
6.37 ms ± 325 µs per loop   this PR
```

Maybe the bigger impact is that fewer objects are created -> fewer objects to garbage collect, which may be important if many indexes are created continiously.